web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts

```diff
@@ -46,74 +52,184 @@ export const useFileTransfer = (
+        name: request.filename,
+        runFileTransfer: abortController =>
+          upload(request.location, file, abortController, {
+            fileRequestId: request.requestID,
+            moderatedSessionId: request.sid,
+          }),
+      });
+    },
+    [currentUser.username, download, filesStore, upload]
+  );
+
+  // handleFileTransferUpdate is called when a FILE_TRANSFER_REQUEST event is received. This is used when
+  // we receive a new file transfer request, or when a request has been updated with an approval but its policy isn't
+  // completely approved yet. An update in this way generally means that the approver array is updated.
+  function handleFileTransferUpdate(data: FileTransferRequest) {
+    setFileTransferRequests(prevstate => {
+      // We receive the same data type when a file transfer request is created and
+      // when an update event happens. Check if we already have this request in our list. If not
+      // in our list, we add it
+      const foundRequest = prevstate.find(
+        ft => ft.requestID === data.requestID
+      );
+      if (!foundRequest) {
+        return [...prevstate, data];
+      } else {
+        return prevstate.map(ft => {
+          if (ft.requestID === data.requestID) {
+            return data;
+          }
+          return ft;
+        });
+      }
+    });
+  }
+
   useEffect(() => {
     // the tty will be init outside of this hook, so we wait until
     // it exists and then attach file transfer handlers to it
-    if (tty) {
-      tty.on(EventType.FILE_TRANSFER_REQUEST, updateFileTransferRequests);
-      tty.on(
-        EventType.FILE_TRANSFER_REQUEST_APPROVE,
-        handleFileTransferApproval
-      );
-      tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
+    if (!tty) {
+      return;
     }
-
+    tty.on(EventType.FILE_TRANSFER_REQUEST, handleFileTransferUpdate);
+    tty.on(EventType.FILE_TRANSFER_REQUEST_APPROVE, handleFileTransferApproval);
+    tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
     return () => {
-      if (tty) {
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST,
-          updateFileTransferRequests
-        );
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST_APPROVE,
-          handleFileTransferApproval
-        );
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST_DENY,
-          handleFileTransferDenied
-        );
+      if (!tty) {
+        return;
       }
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST,
+        handleFileTransferUpdate
+      );
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST_APPROVE,
+        handleFileTransferApproval
+      );
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST_DENY,
+        handleFileTransferDenied
+      );
     };
-  }, [tty]);
+  }, [tty, handleFileTransferDenied, handleFileTransferApproval]);
```

Comment:
Because of how `useSshSession` was originally written, `tty` comes from a ref. React will not re-render the component when `tty` changes so this effect works only "by accident", most probably because the same effect in `useSshSession` which sets `ttyRef.current` also sets `session` which is stored in the state and passed to `useFileTransfer`.

I'm not sure if it's a good practice to store thing in a ref but then pass it around without the ref. Alas, that's how `useSshSession` works on master at the moment.

If you don't feel like refactoring all of this right now (I wouldn't), I'd at least leave a TODO comment about refactoring `useSshSession` to pass `ttyRef` around instead and mentioning that this hook here works only incidentally.